lib/mix/tasks/migrate.ex

```diff
@@ -6,28 +6,110 @@ defmodule Mix.Tasks.Archethic.Migrate do
   alias Archethic.DB.EmbeddedImpl
   alias Archethic.DB.EmbeddedImpl.ChainWriter
 
-  def run(_arg) do
-    version =
-      :archethic
-      |> Application.spec(:vsn)
-      |> List.to_string()
+  @doc """
+  Pre start is triggered before the node is started.
+  In this case we check the last migration version done and run the migration scripts
+  from the last migration done to the last migration script available
 
-    file_path = EmbeddedImpl.db_path() |> ChainWriter.migration_file_path()
+  Pre upgrade is triggered before starting a hot reload
 
-    migration_done? =
-      if File.exists?(file_path) do
-        file_path |> File.read!() |> String.split(";") |> Enum.member?(version)
+  Post upgrade is triggerd after a succesfull hot reload
+  """
+  # Called by migrate.sh scripts
+  def run([function]) when function in ["pre_start", "pre_upgrade", "post_upgrade"] do
+    Application.spec(:archethic, :vsn)
+    |> List.to_string()
+    |> do_run(String.to_existing_atom(function))
+  end
+
+  # Called by MigrateTest
+  def run([function, version])
+      when function in [:pre_start, :pre_upgrade, :post_upgrade] and is_binary(version),
+      do: do_run(version, function)
+
+  def run(_), do: :ok
+
+  defp do_run(current_version, function) do
+    migration_file_path = EmbeddedImpl.db_path() |> ChainWriter.migration_file_path()
+
+    migrations_to_run =
+      if File.exists?(migration_file_path) do
+        read_file(migration_file_path) |> filter_migrations_to_run(function)
```

Comment:
There's something I don't get but that might not be an issue, if there are 4 migrations to catch up, this will do the 4 pre_start one after the other. Is it ok ? Don't you need to do pre_start1 then pre_up1 then post_up1 and only then run pre_start2 then pre_up2, then post_up2 etc. ?

Reply:
The case here is that the pre_start is triggered when the node start, and the pre/post upgrade are triggered when the node is currently running and doing an hot upgrade.
When a node is currenlty running it means that it is on the last version, or it will integrate new version one by one (with the governance transactions).
So for a specific version you can run either pre_start or pre/post upgrade but never both.
And in theory the code in pre_start would be the same as pre/post upgrade